Install Target and Platform Update (#534)
cylc/cylc-doc: src/reference/config/writing-platform-configs.rst

which platforms share install targets. Cylc will then use this information to
make the correct installations on remote platforms, including installation of
files, creation of :cylc:conf:`global.cylc[install][symlink dirs]` and
authentication keys to enable secure communication between platforms.

Note, if missing from configuration, the install target will default to the
platform name. If incorrectly configured, this will cause errors in
:ref:`RemoteInit`.

If you log into one system and see the same files as on another, then these two
platforms will require the same install target in ``global.cylc`` config file.

Example Platform Configurations
-------------------------------

Detailed below are some examples of common platform configurations.

Submit PBS Jobs from Localhost
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- **The** :term:`scheduler` **runs on the** ``localhost`` **platform.**
- **Platforms can share hosts without sharing job runners.**

.. admonition:: Scenario

   You have a cluster where you can submit jobs from the Cylc scheduler host
   using PBS.

The ``localhost`` platform is the Cylc Scheduler host, as configured in

> :cylc:conf:`global.cylc[scheduler][run hosts]available`. This is the host that

Reviewer: Add link to the section added in #542

the workflow will start on.
Our platform ``pbs_cluster`` shares this ``localhost`` host and setting the
install target to ``localhost`` ensures that Cylc knows this platform does not
require remote initialization.

.. code-block:: cylc
   :caption: part of a ``global.cylc`` config file

   [platforms]
       [[localhost]]
           hosts = localhost

>            install target = localhost

Reviewer: This isn't required. I assume you want to include it for clarity?
You could have it commented out? e.g.
```
       # The localhost platform is available by default
       #[[localhost]]
       #    hosts = localhost
       #    install target = localhost
```
Or just remove it and mention it in the paragraph above.

       [[pbs_cluster]]
           hosts = localhost
           job runner = pbs
           install target = localhost

Our Cylc scheduler does not have a job runner defined. Any job submitted to
this ``localhost`` platform will run as a background job. Users can now set 

> :cylc:conf:`flow.cylc[runtime][<namespace>]platform` = pbs_cluster to run pbs jobs.

Reviewer: pbs_cluster needs quoting

- **Platform names can be defined as regular expressions.**

.. admonition:: Scenario

   Everyone in your organization has a computer called ``desktopNNN``,
   all with a file system shared with the scheduler host. Many users
   will want their desktop set up as a platform to run small jobs.

In this scenario, Cylc does not need to install files on the desktop, since
required files which are on the scheduler host will be accessible on the
desktop. From Cylc's point of view, the desktop and scheduler hosts are
considered different platforms but must share an :term:`install target`.
Cylc needs to be told that these platforms share an install target and so we
configure this using the designated configuration item:
:cylc:conf:`global.cylc[platforms][<platform name>]install target`.
Cylc has optional configuration ``[[[meta]]]`` to add a description of the
platform, this may be helpful to use, we will add a platform description to our
desktop platform.
The following platform definition is simplified, taking advantage of defaults
for ``hosts`` and ``install targets``.

.. code-block:: cylc
   :caption: the ``global.cylc`` config file for this scenario could look like:

   [platforms]

>        [[localhost]]

Reviewer: Remove this line?

   because the job ID is only valid on the submission host.

We have set ``retrieve job logs = True``. This will ensure our job logs are
fetched from the ``slurm_cluster`` platform. This setting is recommended for
any remote platform (i.e. where install target is not localhost).

Grouping Platforms
^^^^^^^^^^^^^^^^^^

- **Platform groups allow users to ask for jobs to be run on any
  suitable computer.**

.. admonition:: Scenario

   Extending the example from above, we now wish to set the ``slurm_cluster``
   up such that one host from ``slurm_cluster`` can accept background jobs.
   We would like to group these platforms together so users can set
   :cylc:conf:`flow.cylc[runtime][<namespace>]platform` = supercomputer.

.. code-block:: cylc
   :caption: part of a ``global.cylc`` config file

   [platforms]
       [[slurm_cluster, slurm_cluster_bg]]  # settings that apply to both:
           install target = slurm_cluster
           retrieve job logs = True
       [[slurm_cluster]]
           batch system = slurm
           hosts = login_node_1, login_node_2
       [[slurm_cluster_bg]]
           hosts = login_node_1
       [platform groups]
           [[supercomputer]]

>                platforms = slurm_cluster, slurm_cluster_bg

Reviewer: I don't think it makes sense to have a group mixing slurm & background jobs.
The example I was thinking of is wanting to run a background job on either of the login nodes so the config would be:
```
[platforms]
    [[slurm_cluster, slurm_cluster_bg1, slurm_cluster_bg2]]  # settings that apply to all:
        install target = slurm_cluster
        retrieve job logs = True
    [[slurm_cluster]]
        hosts = login_node_1, login_node_2
        job runner = slurm
    [[slurm_cluster_bg1]]
        hosts = login_node_1
    [[slurm_cluster_bg2]]
        hosts = login_node_2
[platform groups]
    [[slurm_cluster_bg]]
        platforms = slurm_cluster_bg1, slurm_cluster_bg2
```

Group platforms together using the configuration item
:cylc:conf:`global.cylc[platform groups]`. In the above example, the platforms
``slurm_cluster`` and ``slurm_cluster_bg`` both share a file system (
install target = ``slurm_cluster``). We advise caution when grouping platforms
with different install targets as users could encounter a scenario whereby
files (installed by Cylc during the remote initialization process) are

> not available to them.

Reviewer: That's not right (I hope!).
I guess the bit in brackets should be "(created by a previous task using the same platform group)".

:cylc:conf:`global.cylc[install][symlink dirs]`.
The entire workflow directory can be symlinked, using the config item ``run`` 
The following sub-directories  are also available for configuration:

   * log
   * share
   * share/cycle
   * work

These should be configured per :term:`install target`.

For example, to configure workflow ``log`` directories (on the
:term:`scheduler` host) so that they symlink to a different location,
you could write the following in ``global.cylc``:

.. code-block:: cylc

   [install]
       [[symlink dirs]]
           [[[localhost]]]
               log = /somewhere/else

This would result in the following file structure on the Cylc Scheduler:

.. code-block:: none

   ~/cylc-run
   └── myflow
       ├── flow.cylc
       ├── log -> /somewhere/else/cylc-run/myflow/log
       ...

   /somewhere
   └── else
       └── cylc-run
           └── myflow
               └── log
                   ├── flow-config
                   ├── install
                   ...

These ``localhost`` symlinks are created during the cylc install process.
Symlinks for remote install targets are created during :ref:`RemoteInit` following
``cylc play``.

> Symlinks 

Reviewer: Not sure why this line is here?

- Allow a compute cluster with multiple login nodes to be treated as a single
  unit.
- Allow Cylc to elegantly handle failure to communicate with login nodes.
- Reduce the number of ssh connections required for job submission and polling.

.. _Install Targets:

What Are Install Targets?
^^^^^^^^^^^^^^^^^^^^^^^^^

Install targets represent file systems. More than one platform can use the

> same file system. Cylc relies on the site setup file ``global.cylc`` to determine

Reviewer 2: I think we use "config" or "configuration" elsewhere. IMO "setup" is better, but consistency is better still.
```suggestion
same file system. Cylc relies on the site configuration file ``global.cylc`` to determine
```

> which platforms share install targets. Cylc will then use this information to
> make the correct installations on remote platforms, including installation of
> files, creation of :cylc:conf:`global.cylc[install][symlink dirs]` and
> authentication keys to enable secure communication between platforms.

Reviewer 2: Hemmingway says this sentence has a very high reading age. 

```suggestion
which platforms share install targets.

Cylc will setup each remote install target once. During setup it will:

  - Install workflow files
  - Symlink directories
  - Copy authentication keys (to allow secure communication)
```

Our Cylc scheduler does not have a job runner defined. Any job submitted to
this ``localhost`` platform will run as a background job. Users can now set 
:cylc:conf:`flow.cylc[runtime][<namespace>]platform` = ``pbs_cluster`` to run
pbs jobs.

.. note::

   Both ``hosts`` and ``install target`` default to the platform name.

Multiple Platforms Sharing File System with Cylc Scheduler
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- **Platform names can be defined as regular expressions.**

.. admonition:: Scenario

   Everyone in your organization has a computer called ``desktopNNN``,
   all with a file system shared with the scheduler host. Many users
   will want their desktop set up as a platform to run small jobs.

In this scenario, Cylc does not need to install files on the desktop, since
required files which are on the scheduler host will be accessible on the
desktop. From Cylc's point of view, the desktop and scheduler hosts are
considered different platforms but must share an :term:`install target`.
Cylc needs to be told that these platforms share an install target and so we
configure this using the designated configuration item:

> :cylc:conf:`global.cylc[platforms][<platform name>]install target`.
> Cylc has optional configuration ``[[[meta]]]`` to add a description of the
> platform, this may be helpful to use, we will add a platform description to our
> desktop platform.
> The following platform definition is simplified, taking advantage of defaults

Reviewer 2: - I think it's worth saying that it's the platform config (not Cylc) which has the `meta` section, since both workflows and tasks also have a meta section.

- I think paragraph breaks for readability.

- I've tried to make it clearer that metadata is there to help users. (In the workflow and task it can have a functional purpose.)

```suggestion
:cylc:conf:`global.cylc[platforms][<platform name>]install target`.

:cylc:conf:`global.cylc[platforms][<platform name>] has optional configuration 
``[[[meta]]]`` which users can view with ``cylc config --platforms``. We will add
a description designed to help users in this example.

The following platform definition is simplified, taking advantage of defaults
```

   ``^desktop\d\d\d$``.

If a user wants to run a job on their local desktop, e.g. "desktop123", they should
set:

   .. code-block:: cylc

      [runtime]
          [[mytask]]
              platform = desktop123

in their workflow configuration.
If ``[runtime][mytask]platform`` is unset, the job will run on the Cylc
Scheduler host using this default ``localhost`` platform.

Neither platforms will require remote initialization as the ``install target``
is set to ``localhost``.

Cluster with Multiple Login Nodes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- **Platforms with multiple hosts require job runner to be set**
- **Platforms can group multiple hosts together.**

.. admonition:: Scenario

   You have a cluster where users submit jobs to Slurm from
   either of a pair of identical login nodes which share a file system.

.. code-block:: cylc
   :caption: part of a ``global.cylc`` config file

   [platforms]
       [[slurm_cluster]]
           hosts = login_node_1, login_node_2
           job runner = slurm
           retrieve job logs = True

The ``slurm_cluster`` hosts do not share a file system with the scheduler,
therefore ``slurm_cluster`` is a remote platform.

> As the ``install target`` setting for each platform has been omitted, this will

Author: 👍 Changed.

fetched from the ``slurm_cluster`` platform. This setting is recommended for
all remote platforms (i.e. where install target is not localhost).

Grouping Platforms
^^^^^^^^^^^^^^^^^^

- **Platform groups allow users to ask for jobs to be run on any
  suitable computer.**

.. admonition:: Scenario

   Extending the example from above, we now wish to set the ``slurm_cluster``
   up such that ``slurm_cluster`` nodes can accept background jobs.
   We would like to group these background platforms together so users can set
   :cylc:conf:`flow.cylc[runtime][<namespace>]platform` = ``slurm_cluster_bg``.

.. code-block:: cylc
   :caption: part of a ``global.cylc`` config file

   [platforms]
       [[slurm_cluster, slurm_cluster_bg1, slurm_cluster_bg2]]  # settings that apply to all:
           install target = slurm_cluster
           retrieve job logs = True
       [[slurm_cluster]]
           batch system = slurm
           hosts = login_node_1, login_node_2
       [[slurm_cluster_bg1]]
           hosts = login_node_1
       [[slurm_cluster_bg2]]
           hosts = login_node_2
       [platform groups]
           [[slurm_cluster_bg]]

>                platforms = slurm_cluster_bg1, slurm_cluster_bg2

Reviewer: Wrong indentation

Author: This is fixed now.

- Allow a compute cluster with multiple login nodes to be treated as a single
  unit.
- Allow Cylc to elegantly handle failure to communicate with login nodes.
- Configure multiple platforms with the same hosts; for example you can use
  separate platforms to submit jobs to a batch system and to background on

>   ``localhost``.

Reviewer: I don't think this a valid reason why platforms were introduced - this was just as easy before.
Instead we could add:
```
- Reduce the number of ssh connections required for job submission and polling.
```

^^^^^^^^^^^^^^^^^^^^^^^^^

Install targets represent file systems. More than one platform can use the
same file system. Cylc relies on the site setup file ``global.cylc`` to determine
which platforms share install targets. Cylc will then use this information to
make the correct installations on remote platforms, including installation of
files, creation of :cylc:conf:`global.cylc[install][symlink dirs]` and
authentication keys to enable secure communication between platforms.

Note, if missing from configuration, the install target will default to the
platform name. If incorrectly configured, this will cause errors in
:ref:`RemoteInit`.

> To check if two platforms share a file system:

Reviewer: I would remove this bit - I don't think we need to explain how to do this

Create a test file on the first platform:

   .. code-block:: console

      $ ssh platform-A
      $ echo "test file" > cylc-test-file
      $ exit

Check the second platform for the file:

   .. code-block:: console

      $ ssh platform-B
      $ cat cylc-test-file
      $ exit

If the test file exists, then these two platforms share a file system and will
require the same install target in ``global.cylc`` config file.

Example Platform Configurations
-------------------------------

Detailed below are some examples of common platform configurations.

Multiple Platforms Sharing File System with Cylc Scheduler
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> - **The Scheduler Host (Cylc Server) provides a built in localhost platform**

Reviewer: Suggest replacing this with:
```
The :term:`scheduler` runs on the ``localhost`` platform.
```
and add a link to the new section added in #542

- **Platform names can be defined as regular expressions.**

.. admonition:: Scenario

   Everyone in your organization has a computer called ``desktopNNN``,
   all with a file system shared with the scheduler host. Many users
   will want their desktop set up as a platform to run small jobs.

In this scenario, Cylc does not need to install files on the desktop, since
required files which are on the scheduler host will be accessible on the
desktop. From Cylc's point of view, the desktop and scheduler hosts are
considered different platforms but must share an :term:`install target`.
Cylc needs to be told that these platforms share an install target and so we
configure this using the designated configuration item:
:cylc:conf:`global.cylc[platforms][<platform name>]install target`.

.. code-block:: cylc
   :caption: the ``global.cylc`` config file for this scenario could look like:

   [platforms]
       [[localhost]]
           hosts = localhost

>            install target = localhost

Reviewer: These lines are not needed and should be removed.

Reviewer: Ah, I see that they get removed in the text below.
I think it might be simpler to just explain that the localhost  platform does not need to be configured in this case since it is setup by default.

       [[desktop\d\d\d]]
           install target = localhost

The ``localhost`` platform is the Cylc Scheduler host, as configured in
:cylc:conf:`global.cylc[scheduler][run hosts]available`. This is the host that

> the workflow will start on.

Reviewer: Remove these lines - now covered above?

The ``desktop\d\d\d`` platform will set up 1000 platforms, all with the same

> specification and one host per platform.

Reviewer: Suggest:
```
 ``desktop\d\d\d`` is a pattern which defines multiple platforms.
When using a pattern the "hosts" setting must be left unset so that it defaults to the platform name.
This ensures each of the matching platforms is unique.
```

.. note::

   Cylc carries out a "fullmatch" regular expression comparison with the
   the platform name so ``desktop\d\d\d`` is effectively the same as
   ``^desktop\d\d\d$``.

With this config setup, Cylc is aware that these two platforms do not require

> remote installations. Since the job runner is not set, this will default to

Reviewer: Suggest:
```
With this config setup, Cylc is aware that the ``desktop\d\d\d`` platforms share
the same file systems as the scheduler host so do not require remote installation.
```

background.
Cylc has optional configuration ``[[[meta]]]`` to add a description of the
platform, this may be helpful to use, we will add a platform description to our
desktop platform.
In addition, we can take advantage of inbuilt defaults for install target and
hosts,  resulting in a simplified configuration.

.. code-block:: diff

   [platforms]
       [[localhost]]
   -        hosts = localhost
   -        install target = localhost
       [[desktop\d\d\d]]
           install target = localhost
   +       [[[meta]]]
   +           description = "Background job on a desktop system"

If a user wants to run a job on their local desktop, e.g. "desktop123", they should
set:

   .. code-block:: cylc

      [runtime]
          [[mytask]]
              platform = desktop123

in their workflow configuration.
If ``flow.cylc[runtime][mytask]platform`` is unset, the job will run on the Cylc
Scheduler host using this default ``localhost`` platform. It may be appropriate
to allow users to run small jobs on the Cylc Server however more intensive jobs

> should run on a more appropriate platform, as the next scenario will detail.

Reviewer: Cylc can be used to run workflows on large servers with all jobs running locally.
Suggest removing this last sentence.

Cluster with Multiple Login Nodes
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

- **Platforms with multiple hosts require job runner to be set**
- **Platforms can group multiple hosts together.**

.. admonition:: Scenario

>    You have a cluster where users submit to a single Slurm job queue from

Reviewer: "to a single Slurm job queue" -> "jobs to Slurm"
(Slurm can have multiple queues)

   either of a pair of identical login nodes which share a file system.

.. code-block:: cylc
   :caption: part of a ``global.cylc`` config file

   [platforms]

>        [[localhost]]  # Cylc Scheduler

Reviewer: No need for this line?

>        [[spice_cluster]]

Reviewer: "slurm_cluster" would be a more generic name!

Reviewer: (comment applies to the next line)

^^^^^^^^^^^^^^^^^^^^^^^^^

Install targets represent file systems. More than one platform can use the
same file system. Cylc relies on the site setup file ``global.cylc`` to determine
which platforms share install targets. Cylc will then use this information to
make the correct installations on remote platforms, including installation of
files, creation of :cylc:conf:`global.cylc[install][symlink dirs]` and
authentication keys to enable secure communication between platforms.

Note, if missing from configuration, the install target will default to the
platform name. If incorrectly configured, this will cause errors in
:ref:`RemoteInit`.

To check if two platforms share a file system:

Create a test file on the first platform:

   .. code-block:: console

      $ ssh platform-A
      $ echo "test file" > cylc-test-file
      $ exit

Check the second platform for the file:

   .. code-block:: console

      $ ssh platform-B
      $ cat cylc-test-file
      $ exit

> If the test file exists, then these two platforms share a file system and will
> require the same install target in ``global.cylc`` config file.

Reviewer 2: ```suggestion
If the test file exists, then these two platforms share a file system and should
use the same install target in ``global.cylc`` config file.
```

Example Platform Configurations
-------------------------------

.. admonition:: Scenario

>    Cylc scheduler hosts share a file system with a compute cluster.

Reviewer 2: This is almost the same scenario as "Install PBS jobs from localhost"

configure this using the designated configuration item:
:cylc:conf:`global.cylc[platforms][<platform name>]install target`.

.. code-block:: cylc
   :caption: the ``global.cylc`` config file for this scenario could look like:

   [platforms]
       [[localhost]] # cylc-scheduler platform
           hosts = host1, host2
           # unset install target defaults to platform name: localhost
       [[cluster]]
           hosts = host3, host4
           install target = localhost

Cylc is now aware that these two platforms do not require remote installations.

If ``flow.cylc[runtime][mytask]platform`` is unset, the job will run on the Cylc
Scheduler host using this default ``localhost`` platform. It may be appropriate
to allow users to run small jobs on the Cylc Server however more intensive jobs
could be run on the ``cluster``, by setting
``flow.cylc[runtime][mytask]platform = cluster``.

.. admonition:: Scenario

   Mirrored clusters, each with their own file system.

In this case, each cluster would be a platform with its own install target.

.. code-block:: cylc
   :caption: the ``global.cylc`` config file for this scenario could look like:

>    [platforms]
>        [[cluster_1]] # cylc-scheduler platform
>            hosts = host_1, host_2
>            # unset install target defaults to platform name: cluster_1
>        [[cluster_2]]
>            hosts = host_3, host_4
>            # unset install target defaults to platform name: cluster_2

Reviewer 2: I find the word "unset" confusing:
```suggestion
   [platforms]
       [[cluster_1]] # cylc-scheduler platform
           hosts = host_1, host_2
           # install target defaults to platform name: cluster_1
       [[cluster_2]]
           hosts = host_3, host_4
           # install target defaults to platform name: cluster_2
```

or even
```suggestion
   [platforms]
       [[cluster_1]] # cylc-scheduler platform
           hosts = host_1, host_2
           # install target = platform name (cluster_1)
       [[cluster_2]]
           hosts = host_3, host_4
           # install target = platform name (cluster_2)
```

   [platforms]
       [[myhost]]
           hosts = myhost

>            # unset install target defaults to myhost

Reviewer 2: Not sure I like the unset word.
